src/common/subprocess.ts

```diff
@@ -0,0 +1,66 @@
+/** Copyright © 2024-present P0 Security
+
+This file is part of @p0security/cli
+
+@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.
+
+@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
+**/
+import { print2 } from "../drivers/stdio";
+import { AgentArgs } from "../plugins/ssh-agent/types";
+import { spawn, SpawnOptionsWithoutStdio } from "node:child_process";
+
+/** Spawns a subprocess with given command, args, and options.
+ * May write content to its standard input.
+ * Stdout and stderr of the subprocess is printed to stderr in debug mode.
+ * The returned promise resolves with stdout or rejects with stderr of the subprocess. */
+export const asyncSpawn = async (
+  { debug }: AgentArgs,
+  command: string,
+  args?: ReadonlyArray<string>,
+  options?: SpawnOptionsWithoutStdio,
+  writeStdin?: string
+) =>
+  new Promise<string>((resolve, reject) => {
+    const child = spawn(command, args, options);
+
+    let stdout = "";
+    let stderr = "";
+
+    if (writeStdin) {
+      if (!child.stdin) return reject("Child process has no stdin");
+      child.stdin.write(writeStdin);
+    }
+
+    child.stdout.on("data", (data) => {
+      const str = data.toString("utf-8");
+      stdout += str;
```

Comment:
Additionally, this takes a performance hit if there are many `on("data")` calls.

If you instead use a stream or a buffer you eliminate the constant re-allocation of memory here.